meilisearch-core/src/raw_indexer.rs

```diff
@@ -113,52 +141,57 @@ impl<A: AsRef<[u8]>> RawIndexer<A> {
     }
 }
 
-fn index_token<A>(
+fn process_tokens<'a>(tokens: impl Iterator<Item = Token<'a>>) -> impl Iterator<Item = (usize, Token<'a>)> {
+    tokens
+        .scan((0, None), |(offset, sepkind), token| {
+                match token.kind {
+                    TokenKind::Word | TokenKind::StopWord | TokenKind::Any => {
+                        *offset += match *sepkind {
+                            Some(SeparatorKind::Hard) => 8,
+                            Some(SeparatorKind::Soft) => 1,
+                            None => 0,
+                        };
+                        *sepkind = None;
+                    }
+                    TokenKind::Separator(SeparatorKind::Hard) => {
+                        *sepkind = Some(SeparatorKind::Hard);
+                    }
+                    TokenKind::Separator(SeparatorKind::Soft) if sepkind.is_none() => {
+                        *sepkind = Some(SeparatorKind::Soft);
+                    }
+                    _ => (),
+                }
+            Some((*offset, token))
+        })
+    .filter(|(_, t)| t.is_word())
+}
+
+fn index_token(
     token: Token,
+    word_pos: usize,
+    token_pos: usize,
     id: DocumentId,
     indexed_pos: IndexedPos,
     word_limit: usize,
-    stop_words: &fst::Set<A>,
     words_doc_indexes: &mut BTreeMap<Word, Vec<DocIndex>>,
     docs_words: &mut HashMap<DocumentId, Vec<Word>>,
 ) -> bool
-where A: AsRef<[u8]>,
 {
-    if token.index >= word_limit {
+    if token_pos >= word_limit {
         return false;
     }
 
-    let lower = token.word.to_lowercase();
-    let token = Token {
-        word: &lower,
-        ..token
-    };
-
-    if !stop_words.contains(&token.word) {
-        match token_to_docindex(id, indexed_pos, token) {
+    if !token.is_stopword() {
```

Comment:
is it useful? We should have only `words` because of the `filter` in `process_tokens` function.

Reply:
this is a safeguard, and it explains the intention of the code.